### PR TITLE
[TTAHUB-930/986] Use ActivityReportObjective to store cached objective associations

### PR DESCRIPTION
**Pull request by thewatermethod**
## Description of change

This PR builds upon #988 to add linkages from the new ActivityReportObjective* tables to the services that get and save goals for Activity Reports

## How to test

Ensure that the logic in the services makes sense given the data we want to display
Ensure that the basic saving/fetching tasks when editing an AR still work on the frontend

## Issue(s)

* https://ocio-jira.acf.hhs.gov/browse/TTAHUB-986
* https://ocio-jira.acf.hhs.gov/browse/TTAHUB-930

### Every PR

- [ ] Meets issue criteria
- [ ] JIRA ticket status updated
- [ ] Code is meaningfully tested
- [ ] Meets accessibility standards (WCAG 2.1 Levels A, AA)
- [ ] API Documentation updated
- [ ] Boundary diagram updated
- [ ] Logical Data Model updated
- [ ] [Architectural Decision Records](https://adr.github.io/) written for major infrastructure decisions

### Production Deploy

- [ ] Staging smoke test completed

### After merge/deploy

- [ ] Update JIRA ticket status

### CODE REVIEW

File: src/models/activityReport.js [hunk 1]
```diff
@@ -32,6 +32,13 @@ module.exports = (sequelize, DataTypes) => {
       ActivityReport.hasMany(models.NextStep, { foreignKey: 'activityReportId', as: 'specialistNextSteps' });
       ActivityReport.hasMany(models.NextStep, { foreignKey: 'activityReportId', as: 'recipientNextSteps' });
       ActivityReport.hasMany(models.ActivityReportApprover, { foreignKey: 'activityReportId', as: 'approvers', hooks: true });
+      ActivityReport.hasMany(models.ActivityReportGoal, { foreignKey: 'activityReportId', as: 'activityReportGoals' });
+      // ActivityReport.belongsToMany(models.Goal, {
```
Comment (AdamAdHocTeam) [inline src/models/activityReport.js:36]: Do we want this commented out?

File: src/migrations/20220802000000-goal-objective-metadata-caching.js [hunk 2]
```diff
@@ -0,0 +1,380 @@
+module.exports = {
```
[the remark this answers is not in the document]
Comment (thewatermethod) [reply src/migrations/20220802000000-goal-objective-metadata-caching.js:20]: Good idea, very true. I added something but @GarrettEHill should probably check it over for correctness as the author of this migration

File: frontend/src/pages/ActivityReport/Pages/goalsObjectives.js [hunk 3]
```diff
@@ -66,7 +66,17 @@ const GoalsObjectives = ({ reportId }) => {
       try {
         if (isRecipientReport && hasGrants) {
           const fetchedGoals = await getGoals(grantIds);
-          const formattedGoals = fetchedGoals.map((g) => ({ ...g, grantIds }));
+          const formattedGoals = fetchedGoals.map((g) => {
+            // if the goal is on an "old" grant, we should
+            // treat it like a new goal for now
+            let isNew = false;
```
Comment (AdamAdHocTeam) [inline frontend/src/pages/ActivityReport/Pages/goalsObjectives.js:72]: I think this code breaks the retrieval of getting the objectives for existing goals.

see GoalForm.js line 98 "if (!goal.isNew && goal.goalIds) " the first part of this IF is never true for a goal with objectives. So it never attempts to retrieve the goal objectives from the db.

Comment (AdamAdHocTeam) [reply frontend/src/pages/ActivityReport/Pages/goalsObjectives.js:72]: Let me know once this is fixed and I will continue testing.

Comment (thewatermethod) [reply frontend/src/pages/ActivityReport/Pages/goalsObjectives.js:72]: This issue has been addressed

File: src/lib/orderGoalsBy.js [hunk 4]
```diff
@@ -1,20 +1,32 @@
...
+ * @param {String} sortBy
+ * @param {String} sortDir
+ * @returns Array[] of shape [sequelize.col(), String sortDir]
+ */
+
+// note that this can only be used when a) Status Sort is a aliased SQL function and
+// b) it's the first aliased variable due to the way Sequelize minifies arrays
+
+// storing this in a constant so it looks a little less magical/stupid
+export const STATUS_SORT = '_0';
```
Comment (GarrettEHill) [inline src/lib/orderGoalsBy.js:15]: What is this?

Comment (thewatermethod) [reply src/lib/orderGoalsBy.js:15]: that's the aliased table name

File: src/models/activityReport.js (hunk 1 above)
Comment (GarrettEHill) [reply src/models/activityReport.js:36]: I think is was breaking something, we can try uncommenting it.

File: src/models/hooks/activityReport.js [hunk 5]
```diff
@@ -318,6 +318,60 @@ const automaticStatusChangeOnAprovalForGoals = async (sequelize, instance, optio
...
+      include: [{
+        model: sequelize.models.ActivityReport,
+        as: 'activityReports',
+        required: true,
+        where: { id: instance.id },
+      }],
+      transaction: options.transaction,
+    });
+
+    // Update Goal status to 'In Progress'.
```
Comment (GarrettEHill) [inline src/models/hooks/activityReport.js:360]:
```suggestion
    // Update Objective status to 'In Progress'.
```

File: src/services/goalServices/goals.test.js [hunk 6]
```diff
@@ -49,7 +58,7 @@ describe('Goals DB service', () => {
       Objective.create = jest.fn().mockResolvedValue({ id: 1 });
       Objective.findOrCreate = jest.fn().mockResolvedValue([{ id: 1 }]);
       Objective.update = jest.fn().mockResolvedValue({ id: 1 });
-      Objective.findByPk = jest.fn().mockResolvedValue({ update: existingObjectiveUpdate });
+      Objective.findByPk = jest.fn().mockResolvedValue({ id: 1, update: existingObjectiveUpdate });
```
Comment (GarrettEHill) [inline src/services/goalServices/goals.test.js:61]: This one took me 2 days to figure out

**Review by GarrettEHill — APPROVED** [review]

File: src/services/goalServices/goals.test.js (hunk 6 above)
Comment (thewatermethod) [reply src/services/goalServices/goals.test.js:61]: 🤣

File: src/models/activityReport.js (hunk 1 above)
Comment (thewatermethod) [reply src/models/activityReport.js]: I lied, figured it out

File: src/migrations/20220802000000-goal-objective-metadata-caching.js (hunk 2 above)
[the remark this answers is not in the document]
Comment (thewatermethod) [reply src/migrations/20220802000000-goal-objective-metadata-caching.js:20]: as the instantiator* of this migration

**Review by AdamAdHocTeam — COMMENTED** [review]
I think there is an issue with 'auto save' putting the objective 'Topics' into a bad state.

This occurred twice for me when attempting to enter a complex goal set.

1. Create a new recipient AR
2. Create a new goal and objective
3. Click 'Save Goal'
4. Click 'Add new goal'
5. Enter a a name and topics
6. Allow the 'auto save' to occur
7. Complete the objective and save

You should get below server error:

image

image

**Comment by AdamAdHocTeam** [comment]
> 

Disregard my last comment I think it was because I had out of date code. :-)

File: src/services/goals.js [hunk 7]
```diff
@@ -984,21 +984,56 @@ export async function goalsForGrants(grantIds) {
...
+    }),
+    ActivityReportObjectiveResource.destroy({
+      where: {
+        activityReportObjectiveId: activityReportObjectiveIdsToRemove,
+      },
+    }),
+    ActivityReportObjectiveTopic.destroy({
+      where: {
+        activityReportObjectiveId: activityReportObjectiveIdsToRemove,
+      },
```
Comment (GarrettEHill) [inline src/services/goals.js:1002]: Should this be in the other file with the rest of the caching code?

Comment (thewatermethod) [reply src/services/goals.js]: sure, can be

**Review by AdamAdHocTeam — COMMENTED** [review]
If I add files to the objective then save the goal and edit the UI shows duplicate files.

image

Resources are also disappearing on edit 

image

**Review by AdamAdHocTeam — COMMENTED** [review]
Not sure if this is a new jira ticket:

Create a AR with two recipients add one goal and two objectives and save. Then add an additional recipient and save. Note that the objectives table is missing an entry.

**Review by AdamAdHocTeam — APPROVED** [review]
Looks good.

**Pull request merged**